docs/source/theano.rst

```diff
@@ -0,0 +1,216 @@
+
+    func = theano.function([a, x, y], [out])
+
+We can call this function with actual arrays as many times as we want::
+
+    a_val = 1.2
+    x_vals = np.random.randn(10)
+    y_vals = np.random.randn(10)
+
+    out = func(a_val, x_vals, y_vals)
+
+For the most part the symbolic Theano variables can be operated on
+like NumPy arrays. Most NumPy functions are available in `theano.tensor`
+(which is typically imported as `tt`). A lot of linear algebra operations
+can be found in `tt.nlinalg` and `tt.slinalg` (the NumPy and SciPy
+operations respectively). Some support for sparse matrices is available
+in `theano.sparse`. For a detailed overview of available operations,
+see `the theano api docs <http://deeplearning.net/software/theano/library/tensor/index.html>`_.
+
+A notable exception where theano variables do *not* behave like
+NumPy arrays are operations involving conditional execution:
+
+Code like this won't work as expected::
+
+    a = tt.vector('a')
+    if (a > 0).all():
+        b = tt.sqrt(a)
+    else:
+        b = -a
+
+`(a > 0).all()` isn't actually a boolean as it would be in NumPy, but
+still a symbolic variable. Python will convert this object to a boolean
+and according to the rules for this conversion, things that aren't empty
+containers or zero are converted to `True`. So the code is equivalent
+to this::
+
+    a = tt.vector('a')
+    b = tt.sqrt(a)
+
+To get the desired behaviour, we can use `tt.switch`::
+
+    a = tt.vector('a')
+    b = tt.switch((a > 0).all(), tt.sqrt(a), -a)
+
+Indexing also works similarly to NumPy::
+
+    a = tt.vector('a')
+    # Access the 10th element. This will fail when a function build
+    # from this expression is executed with an array that is too short.
+    b = a[10]
+
+    # Extract a subvector
+    b = a[[1, 2, 10]]
+
+Changing elements of an array is possible using `tt.set_subtensor`::
+
+    a = tt.vector('a')
+    b = tt.set_subtensor(a[:10], 1)
+
+    # is roughly equivalent to this (although theano avoids
+    # the copy if `a` isn't used anymore)
+    a = np.random.randn(10)
+    b = a.copy()
+    b[:10] = 1
+
+How PyMC3 uses Theano
+=====================
+
+Now that we have a basic understanding of Theano we can look at what
+happens if we define a PyMC3 model. Let's look at a simple example::
+
+    true_mu = 0.1
+    data = true_mu + np.random.randn(50)
+
+    with pm.Model() as model:
+        mu = pm.Normal('mu', mu=0, sd=1)
+        y = pm.Normal('y', mu=mu, sd=1, observed=data)
+
+In this model we define two variables: `mu` and `y`. The first is
+a free variable that we want to infer, the second is an observed
+variable. To sample from the posterior we need to build the function
+
+.. math::
+
+   \log P(μ|y) + C = \log P(y|μ) + \log P(μ) =: \text{logp}(μ)\\
+
+where with the normal likelihood :math:`N(x|μ,σ^2)`
+
+.. math::
+
+    \text{logp}\colon \mathbb{R} \to \mathbb{R}\\
+    μ \mapsto \log N(μ|0, 1) + \log N(y|0, 1),
```

Comment:
This might be a typo, it should be 

`μ \mapsto \log N(μ|0, 1) + \log N(y|μ, 1),`

right?